javalib/src/main/scala/java/util/Base64.scala

```diff
@@ -0,0 +1,390 @@
+
+  def getUrlEncoder: Encoder = urlSafeEncoder
+
+  // --------------------------------------------------------------------------
+
+  class Decoder private[Base64](table: IntTable, ignoreInvalid: Boolean) {
+
+    def decode(src: Array[Byte]): Array[Byte] = {
+      val dst = ByteBuffer.allocate(dstMaxLength(src.length))
+      doDecode(ByteBuffer.wrap(src), dst)
+      val result = new Array[Byte](dst.position)
+      dst.flip()
+      dst.get(result)
+      result
+    }
+
+    def decode(src: Array[Byte], dst: Array[Byte]): Int = {
+      val dstBb = ByteBuffer.allocate(dstMaxLength(src.length))
+      val written = doDecode(ByteBuffer.wrap(src), dstBb)
+      if(written > dst.length)
+        throw new IllegalArgumentException("Output byte array is too small for decoding all input bytes")
+      dstBb.flip()
+      dstBb.get(dst)
+      written
+    }
+
+    def decode(buffer: ByteBuffer): ByteBuffer = {
+      val result = ByteBuffer.allocate(dstMaxLength(buffer.remaining()))
+      doDecode(buffer, result)
+      result.flip()
+      result
+    }
+
+    def decode(src: String): Array[Byte] = decode(src.getBytes(StandardCharsets.ISO_8859_1))
+
+    def wrap(is: InputStream): InputStream = new DecodingInputStream(is, table, ignoreInvalid)
+
+    private def doDecode(src: ByteBuffer, dst: ByteBuffer): Int = {
+      val srcBuffer = IntBuffer.allocate(4)
+
+      @inline
+      def inputData(): Unit = {
+        srcBuffer.rewind()
+        var shift = 18
+        var i = 0
+        while(srcBuffer.hasRemaining) {
+          i |= (srcBuffer.get() << shift)
+          shift -= 6
+        }
+        if(shift == 12) throw new IllegalArgumentException("Last unit does not have enough valid bits")
+        if(shift <= 6) dst.put((i >> 16).toByte)
+        if(shift <= 0) dst.put((i >> 8).toByte)
+        if(shift <= -6) dst.put(i.toByte)
+        srcBuffer.clear()
+      }
+
+      @inline
+      def iterate(): Unit = {
+        while (src.hasRemaining) {
+          if (srcBuffer.hasRemaining) {
+            val int = src.get() & 0xff
+            table(int) match {
+              case -2 => return
+              case -1 if ignoreInvalid =>
+              case -1 => throw new IllegalArgumentException("Illegal base64 character " + Integer.toString(int, 16))
+              case i => srcBuffer.put(i)
+            }
+          } else inputData()
+        }
+      }
+      iterate()
+      // end or padding
+      srcBuffer.flip()
+      inputData()
+      while(src.hasRemaining) {
+        val int = src.get() & 0xff
+        val value = table(int)
+        if (value != -2 && (!ignoreInvalid || value > 0))
+          throw new IllegalArgumentException(s"Input byte array has incorrect ending byte at $int")
+      }
+      dst.position()
+    }
+
+    private def dstMaxLength(srcLength: Int): Int = (srcLength+3) / 4 * 3
+
+  }
+
+  private class DecodingInputStream(in: InputStream, table: IntTable, ignoreInvalid: Boolean) extends InputStream {
```

Comment:
I recommend you extend `FilterInputStream`. It gives you some correct default implementations.